crates/extensions/c8y_auth_proxy/src/server.rs

```diff
@@ -135,23 +165,131 @@ impl FromRef<AppState> for SharedTokenManager {
+    match message {
+        In::Text(t) => Out::Text(t),
+        In::Binary(t) => Out::Binary(t),
+        In::Ping(t) => Out::Ping(t),
+        In::Pong(t) => Out::Pong(t),
+        In::Close(Some(InCf { code, reason })) => Out::Close(Some(OutCf {
+            code: code.into(),
+            reason,
+        })),
+        In::Close(None) => Out::Close(None),
+    }
+}
+
+fn tungstenite_to_axum(
+    message: tokio_tungstenite::tungstenite::Message,
+) -> axum::extract::ws::Message {
+    use axum::extract::ws::CloseFrame as OutCf;
+    use axum::extract::ws::Message as Out;
+    use tokio_tungstenite::tungstenite::protocol::frame::CloseFrame as InCf;
+    use tokio_tungstenite::tungstenite::Message as In;
+    match message {
+        In::Text(t) => Out::Text(t),
+        In::Binary(t) => Out::Binary(t),
+        In::Ping(t) => Out::Ping(t),
+        In::Pong(t) => Out::Pong(t),
+        In::Close(Some(InCf { code, reason })) => Out::Close(Some(OutCf {
+            code: code.into(),
+            reason,
+        })),
+        In::Close(None) => Out::Close(None),
+        In::Frame(_) => unreachable!("This function is only called when reading a message"),
+    }
+}
+
+async fn connect_to_websocket(
+    token: &str,
+    headers: &HeaderMap<HeaderValue>,
+    uri: &str,
+) -> Result<WebSocketStream<MaybeTlsStream<TcpStream>>, tokio_tungstenite::tungstenite::Error> {
+    let mut req = Request::builder();
+    for (name, value) in headers {
+        req = req.header(name.as_str(), value);
+    }
+    req = req.header("Authorization", format!("Bearer {token}"));
+    let req = req
+        .uri(uri)
+        .body(())
+        .expect("Builder should always work as the headers are copied from a previous request, so must be valid");
+    tokio_tungstenite::connect_async(req)
+        .await
+        .map(|(c8y, _)| c8y)
+}
 
-impl fmt::Display for TargetHost {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.0.fmt(f)
+async fn proxy_ws(
+    ws: WebSocket,
+    host: TargetHost,
+    retrieve_token: State<SharedTokenManager>,
+    headers: HeaderMap<HeaderValue>,
+    path: String,
+) {
+    let uri = format!("{}/{path}", host.ws);
+    let mut token = retrieve_token.not_matching(None).await;
+    let c8y = match connect_to_websocket(&token, &headers, &uri).await {
+        Ok(c8y) => Ok(c8y),
+        Err(_) => {
+            token = retrieve_token.not_matching(Some(&token)).await;
+            match connect_to_websocket(&token, &headers, &uri).await {
+                Ok(c8y) => Ok(c8y),
+                Err(e) => Err(anyhow::Error::from(e)),
+            }
+        }
+    }
+    .expect("Error connecting to proxied websocket");
+    let (mut to_c8y, mut from_c8y) = c8y.split();
+    let (mut to_client, mut from_client) = ws.split();
+    let mut client_to_c8y = tokio::spawn(async move {
+        while let Some(Ok(msg)) = from_client.next().await {
+            to_c8y.send(axum_to_tungstenite(msg)).await.unwrap();
```

Comment:
Any error on forwarding a packet should be returned by `client_to_c8y`.

Reply:
I don't quite understand the value of doing this. It makes failing here more difficult, and it makes handling the error more complicated as we (a) should handle the JoinError (the panic case) anyway and (b) have to remember to handle the inner error additionally to that.